lib/eclipse/EclipseState/Schedule/Well.cpp

```diff
@@ -574,4 +574,10 @@ namespace Opm {
     bool Well::hasEvent(uint64_t eventMask, size_t reportStep) const {
         return this->events.hasEvent( eventMask , reportStep );
     }
+
+
+    void Well::filter_completions(const EclipseGrid& grid) {
+        for (auto& completions : m_completions)
```

Comment:
This was the only part I misunderstood at first: I did not realize that this was the loop over time steps! While I think the addition of begin() and end() to DynamicState is fine, it is not obvious to the reader (well, me at least...) if we are looping over times first and completions second (correct here, since we have DynamicState<CompletionSet>), or the other way around (if a CompletionSet had been a vector of DynamicState<Completion>).

Reply:
Just an idea to make it more obvious: instead of providing begin() and end(), so you can use the DynamicState as a range, it could provide a timeRange() or allTimes() or similarly named functions that returned the iterator range. Then you would write the above as
```
for (auto& completions : m_completions.allTimes())
```
What do you think?

Reply:
I think that was a good idea - what would the implementation of the `DynamicState::allTimes()` method look like? Return a pair of iterators?

Reply:
I am not sure if pair<iterator,iterator> will be treated as a range, if so then that is the best. Worst case you'll have to use or import boost::iterator_range, which is what I have used up to now. There seems to be no std:: replacement for it yet, I did a brief search.

Reply:
OK - I liked your idea; but I did not venture into boost for this. Have added a comment though.

Reply:
That's fine. The comment makes it very clear what is going on.